doc/kernel/usermode/syscalls.rst

```diff
@@ -414,7 +414,7 @@ bytes processed. This too should use a stack copy:
 
         Z_OOPS(z_user_from_copy(&size, size_ptr, sizeof(size));
         ret = z_impl_in_out_syscall(&size);
-        *size_ptr = size;
+        Z_OOPS(z_user_to_copy(size_ptr, &size, sizeof(size)));
```

Comment:
Agree that this check is going to be needed in robust code.  But I do note that there's a Z_SYSCALL_MEMORY_WRITE() validation API that is probably a better fit for memory like this integer that is naturally used directly.  So this example still doesn't look like "best practice" to me.

Reply:
The example with `Z_SYSCALL_MEMORY_WRITE()` would be:
```
int z_vrfy_in_out_syscall(size_t *size_ptr)
{
    size_t size;
    int ret;
    Z_OOPS(Z_SYSCALL_MEMORY_WRITE(size_ptr, sizeof(*size_ptr)));
    size = *size_ptr;
    ret = z_impl_in_out_syscall(&size);
    *size_ptr = size;
    return ret;
}
```

The advantages would be:
- (slight) performance improvement due to only a single memory access check

The disadvantages are:
- loss of clarity regarding when the data crosses the privileged memory domain (`size` on stack) and usermode domain (`*size_ptr`), i.e. lack of clear usermode vs supervisor memory separation
- (arguable, but strongly related to first disadvantage) it is tempting to get rid of the copy and pass the `size_ptr` instead of `&size` to implementation function. Note that this should be avoided as noted two paragraphs above this example.

I don't think `Z_SYSCALL_MEMORY_WRITE()` is a better fit here. I strongly favor correct and clear (to humans) code over micro-optimized code that most likely don't have to be optimized.